docs(core) - docs for tooltip added (#452)
konturio/disaster-ninja-fe: src/core/tooltips/rfc.md

## Tooltip Service

This service allows us to show tooltips in different places of our application.
Underhood it uses `@konturio/ui-kit` tooltip component for show tooltip in theme style

Tooltip:

- Shows by hover-in and hides by hover-out events (after delay)
- Not interactive
- Can contain only text

## How to use

We assume that all cases of using a tooltip can be reduced to two cases

### We have an element around which a tooltip should be shown

In that case you need to import special components and use them

```jsx
import { Tooltip, TooltipTrigger, TooltipContent } from '~core/tooltip';

function Sidebar() {
  return (
    <nav>

>       <Tooltip>

Reviewer: If we need text-only content, we can use a simpler design

```tsx
<Tooltip content="string-only-content">
  {/* Where to show */}
  <button>Some button with tip</button>
</Tooltip>
```

The current design is more compatible with `Popup` which can contain any content

Author: It's better to keep like this.
I can imagine case when you want pass <Trans>  component, or <Markdown> to tootip content.
Also in feature we probably add settings for content and trigger, and it will be cool that we keep it separately

>       <Tooltip>

Reviewer: According to the design, the tooltip can be sized, regular and small, we need some prop for the sizing
<img width="924" alt="image" src="https://user-images.githubusercontent.com/14041229/220108365-09756ff7-ac41-4212-a014-2fe03ad1b8e8.png">

>       <Tooltip>

Reviewer: The tooltip open status will be uncontrolled by default, it will be useful to have an `open` prop for controlled usage.

Author: For what case you need that?
In case you need to open/close it manually you want to use service directly as in example with map

Reviewer: For example, show a tooltip on some component when hovering some another

Author: Sounds like broken ux : )
But if you really need that you still can do it:
```ts
triggerElement.addEventListener('pointer', (e) => {
  tooltip.show(
      ElementWhereToShow,
      'Some text'
    );
})
  ```

        {/* Where to show */}
        <TooltipTrigger>
          <button>Some button with tip</button>
        </TooltipTrigger>
        {/* What to show */}
        <TooltipContent>Text hint</TooltipContent>
      </Tooltip>
    </nav>
  );
}
```

### We have some coordinates on the screen where tooltip should be shown

In that case you need to use imperative api.

For example, let's show countries under cursor

```ts
import { tooltips } from '~core/tooltips';

let currentTooltipId: number | null = null;

map.on('click', (event) => {
  const bbox = [
    [e.point.x - 5, e.point.y - 5],
    [e.point.x + 5, e.point.y + 5],
  ];

  const selectedFeatures = map.queryRenderedFeatures(bbox, {
    layers: ['counties'],
  });

  // Close previous tooltip
  if (currentTooltipId !== null) {
    tooltips.closeTooltip(currentTooltipId);
    currentTooltipId = null;
  }

  // Show new tooltip

>   currentTooltipId = tooltips.showTooltip(

Reviewer 2: so we won't use mapbox tooltips on map?

Reviewer: I think yes, all app tooltips will work the same way

Author: @propakov 
I think it would be easier to maintain one single tooltip service. Is there a reason why mapbox tooltips are better?

Reviewer 3: Are you talking about tooltips or map popups really? Map popups have different behavior from tooltips.

Author: Map have "popups".
It can be used for showing tooltips.
https://docs.mapbox.com/mapbox-gl-js/example/popup-on-hover/

Or classical popups 
https://docs.mapbox.com/mapbox-gl-js/example/popup-on-click/

Reviewer 2: they just sticks to map position, and draw on canvas, so that's the difference

Reviewer 2: agree that maintaining single service is easier if you are ok to go without sticking to the map

    // Where to show
    { x: event.point.x, y: event.point.y },
    // What to show
    selectedFeatures.map((f) => f.properties.countryName).join(' ,'),
  );
});
```

### How service works

```ts
interface TooltipService {

>   showTooltip: ({ x: number, y: number }, content: string) => number);

Reviewer: In some cases it will be useful to pass `element` to `showTooltip`, for example, VisTimeline has a hover callback that provides an `element` arg

## Tooltip Service

> This service allow us show tooltips in different places of our application.

Reviewer 4: ```suggestion
This service allows us to show tooltips in different places of our application.
```

> Underhood it uses `@konturio/ui-kit` tooltip component for show tooltip in theme style

Reviewer 4: "to show" or "for showing"

Also i think all ui-kit components are style-compatible, no?

Author: > Also i think all ui-kit components are style-compatible, no?

Sure. this text sounds like it doesn't?

Reviewer 4: yep

Tooltip:

> - shows by hover-in and hides by hover-out events (after delay)

Reviewer 2: ```suggestion
- Shows by hover-in and hides by hover-out events (after delay)
```

- Not interactive

> - Can contain only text

Reviewer 4: what about rich text\markdown?

## How to use

We assume that all cases of using a tooltip can be reduced to two cases

### We have an element around which a tooltip should be shown

> In that case you need import special components and use them

Reviewer 2: ```suggestion
In that case you need to import special components and use them
```

```jsx
import { Tooltip, TooltipTrigger, TooltipContent } from '~core/tooltip';

function Sidebar() {
  return (
    <nav>

>       <Tooltip>

Reviewer 4: Looks like TooltipContext or TooltipWrap or smth

Reviewer 4: Not subject of this PR i guess

Reviewer: Less words === Simpler

Reviewer 4: Less words === more interpretations 😄

Author: I'am not see any way how it can be intepretated here in wrong way

>         {/* Where show */}

Reviewer 4: ```suggestion
        {/* Where to show */}
```

        <TooltipTrigger>
          <button>Some button with tip</button>
        </TooltipTrigger>

>         {/* What show */}

Reviewer 4: ```suggestion
        {/* What to show */}
```

        <TooltipContent>Text hint</TooltipContent>
      </Tooltip>
    </nav>
  );
}
```

### We have some coordinates on the screen where tooltip should be shown

In that case you need to use imperative api.

For example, let's show countries under cursor

```ts
import { tooltips } from '~core/tooltips';

let currentTooltipId: number | null = null;

map.on('click', (event) => {
  const bbox = [
    [e.point.x - 5, e.point.y - 5],
    [e.point.x + 5, e.point.y + 5],
  ];

  const selectedFeatures = map.queryRenderedFeatures(bbox, {
    layers: ['counties'],
  });

  // Close previous tooltip
  if (currentTooltipId !== null) {
    tooltips.closeTooltip(currentTooltipId);
    currentTooltipId = null;
  }

  // Show new tooltip
  currentTooltipId = tooltips.showTooltip(

>     // Where show
>     { x: event.point.x, y: event.point.y },
>     // What show

Reviewer 2: ```suggestion
    // Where to show
    { x: event.point.x, y: event.point.y },
    // What to show
```

    selectedFeatures.map((f) => f.properties.countryName).join(' ,'),
  );
});
```

### How service works

```ts
interface TooltipService {

>   init(mountPoint: Element) => void;

Author: ```suggestion
```
library use body for automatically create container
https://github.dev/floating-ui/floating-ui

> interface TooltipService {
>   init(mountPoint: Element) => void;
>   showTooltip: ({ x: number, y: number }, content: string) => number);
>   updateTooltip: ({ x: number, y: number }, content: string) => number);

Reviewer 4: ```suggestion
type TooltipId = number;

interface TooltipService {
  init(mountPoint: Element) => void;
  showTooltip: ({ x: number, y: number }, content: string) => TooltipId);
  updateTooltip: ({ x: number, y: number }, content: string) => TooltipId);
```
